### ISSUE TITLE
Istio Sidecars adding large amounts of latency

### ISSUE THREAD

**Issue by bryan-kc** [istio/istio#33283]
(**NOTE**: This is used to report product bugs:
  To report a security vulnerability, please visit <https://istio.io/about/security-vulnerabilities>
  To ask questions about how to use Istio, please visit <https://discuss.istio.io>)

**Bug description**
I've been working on rolling out Istio onto a production setup. We have Istio sidecars with namespace injection. We are still using ALB Ingress in this case. When the sidecars were rolled out we noticed a large increase in p95, p99, p99.9 latency coming from the ALB's reported metrics. These increases of latency are on the order of seconds, which is quite concerning.

[ ] Docs
[X] Installation
[X] Networking
[X] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure
[ ] Upgrade

**Expected behavior**
Sidecar injection should only add minimal amounts of latency to our installation.

**Steps to reproduce the bug**
Create a GraphQL application with an ALB Ingress. Inject Istio Sidecars. Monitor Network performance from Cloudwatch within the AWS console.

![istio](https://user-images.githubusercontent.com/12261570/120833374-92561700-c516-11eb-88ea-54802d441449.png)

This dashboard we have is plotting simple HTTP requests which aren't expected to increase in duration much. We also have websockets for this application but they did not appear to be impacted by the installation.

**Version** (include the output of `istioctl version --remote` and `kubectl version --short` and `helm version --short` if you used Helm)
```
client version: 1.10.0
control plane version: 1.9.1
data plane version: 1.9.1 (65 proxies)
```

**How was Istio installed?**
Istio operator install with the following manifest

```yaml
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
  namespace: istio-system
  name: istiocontrolplane
spec:
  values:
    telemetry:
      v2:
        prometheus:
          configOverride:
            outboundSidecar:
              debug: false
              stat_prefix: istio
              metrics:
                - name: request_duration_milliseconds
                  dimensions:
                    cache_status: response.headers['X-Proxy-Cache']
  profile: default
  meshConfig:
    accessLogFile: /dev/stdout
    enableTracing: true
    defaultConfig:
      tracing:
        sampling: 100
        zipkin:
          address: opentelemetry-collector.opentelemetry-collector:9411
  components:
    pilot:
      k8s:
        hpaSpec:
          minReplicas: 2
        replicaCount: 2
```

**Environment where the bug was observed (cloud vendor, OS, etc)**
AWS EKS
```
Server Version: version.Info{Major:"1", Minor:"16+", GitVersion:"v1.16.15-eks-ad4801", GitCommit:"ad4801fd44fe0f125c8d13f1b1d4827e8884476d", GitTreeState:"clean", BuildDate:"2020-10-20T23:27:12Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}
```

Additionally, please consider running `istioctl bug-report` and attach the generated cluster-state tarball to this issue.
Refer [cluster state archive](http://istio.io/help/bugs/#generating-a-cluster-state-archive) for more details.

**Comment by bianpengyuan** [comment]
The graph says rolling out of Istio increases latency by hundreds of seconds? That does not seem to be right. Does this agree with Istio stats?

**Comment by bryan-kc** [comment]
I have some data from Kiali that suggests this is the case. We also noticed that persistent connections may have been killed at the 5 minute mark, but we are unsure if this was related.

Below is two panels of response time for the same service. Slicing the data into response flags could reveal similar data. 
![Screen Shot 2021-06-04 at 12 56 01 PM](https://user-images.githubusercontent.com/12261570/120856440-be809080-c534-11eb-9a05-6ec6907edd81.png)
![Screen Shot 2021-06-04 at 12 56 33 PM](https://user-images.githubusercontent.com/12261570/120856443-bfb1bd80-c534-11eb-913d-f7228086324d.png)

**Comment by bianpengyuan** [comment]
> We also noticed that persistent connections may have been killed at the 5 minute mark, but we are unsure if this was related.

That seems to be related, because the p99 with UC and UF are at 300s? I am not sure whether Kiali uses server stats or client stats, but have you compared client vs server istio stats latency? Is it possible that the application hangs and does not response properly? Envoy has an native stats `upstream_rq_time` which measures time spent at upstream application and can be exposed by https://istio.io/latest/docs/ops/configuration/telemetry/envoy-stats/ which should give an idea about whether envoy thinks the upstream app that spends all 300s.

**Comment by bryan-kc** [comment]
Kiali was using client stats in this case. I took a look at envoy stats and can retrieve the following. I couldn't get `upstream_rq_time` -- it didn't seem to be exposed.

Here's some stats I retrieved. Note that I am obfuscating naming of our services.
```
cluster.xds-grpc.upstream_cx_connect_ms: P0(nan,0.0) P25(nan,0.0) P50(nan,0.0) P75(nan,0.0) P90(nan,1.08) P95(nan,7.04) P99(nan,7.088) P99.5(nan,7.094) P99.9(nan,7.0988) P100(nan,7.1)
cluster.xds-grpc.upstream_cx_length_ms: P0(nan,1600000.0) P25(nan,1691666.6666666667) P50(nan,1812500.0) P75(nan,1881250.0) P90(nan,1945000.0) P95(nan,1972500.0) P99(nan,1994500.0) P99.5(nan,1997250.0) P99.9(nan,1999450.0) P100(nan,2000000.0)

reporter=.=destination;.;source_workload=.=unknown;.;source_workload_namespace=.=unknown;.;source_principal=.=unknown;.;source_app=.=unknown;.;source_version=.=unknown;.;source_canonical_service=.=unknown;.;source_canonical_revision=.=latest;.;destination_workload=.=apigw;.;destination_workload_namespace=.=project-apigw-integration;.;destination_principal=.=unknown;.;destination_app=.=apigw;.;destination_version=.=unknown;.;destination_service=.=apigw.project-apigw-integration.svc.cluster.local;.;destination_service_name=.=apigw;.;destination_service_namespace=.=project-apigw-integration;.;destination_canonical_service=.=apigw;.;destination_canonical_revision=.=latest;.;request_protocol=.=http;.;response_flags=.=-;.;connection_security_policy=.=none;.;response_code=.=200;.;grpc_response_status=.=;.;destination_cluster=.=Kubernetes;.;source_cluster=.=unknown;.;istio_request_duration_milliseconds: P0(0.0,0.0) P25(0.0,1.0691546762589927) P50(1.025,54.2) P75(1.0625,200.27472527472528) P90(1.085,363.5000000000002) P95(1.0925,1110.3883495145628) P99(1.0985,1486.999999999989) P99.5(1.09925,3086.999999999989) P99.9(1.09985,5777.399999999989) P100(1.1,11000.0)
reporter=.=destination;.;source_workload=.=unknown;.;source_workload_namespace=.=unknown;.;source_principal=.=unknown;.;source_app=.=unknown;.;source_version=.=unknown;.;source_canonical_service=.=unknown;.;source_canonical_revision=.=latest;.;destination_workload=.=apigw;.;destination_workload_namespace=.=project-apigw-integration;.;destination_principal=.=unknown;.;destination_app=.=apigw;.;destination_version=.=unknown;.;destination_service=.=apigw.project-apigw-integration.svc.cluster.local;.;destination_service_name=.=apigw;.;destination_service_namespace=.=project-apigw-integration;.;destination_canonical_service=.=apigw;.;destination_canonical_revision=.=latest;.;request_protocol=.=http;.;response_flags=.=-;.;connection_security_policy=.=none;.;response_code=.=400;.;grpc_response_status=.=;.;destination_cluster=.=Kubernetes;.;source_cluster=.=unknown;.;istio_request_duration_milliseconds: P0(nan,7.0) P25(nan,7.05) P50(nan,7.1) P75(nan,33.5) P90(nan,33.8) P95(nan,33.9) P99(nan,33.98) P99.5(nan,33.99) P99.9(nan,33.998) P100(nan,34.0)
reporter=.=destination;.;source_workload=.=unknown;.;source_workload_namespace=.=unknown;.;source_principal=.=unknown;.;source_app=.=unknown;.;source_version=.=unknown;.;source_canonical_service=.=unknown;.;source_canonical_revision=.=latest;.;destination_workload=.=apigw;.;destination_workload_namespace=.=project-apigw-integration;.;destination_principal=.=unknown;.;destination_app=.=apigw;.;destination_version=.=unknown;.;destination_service=.=apigw.project-apigw-integration.svc.cluster.local;.;destination_service_name=.=apigw;.;destination_service_namespace=.=project-apigw-integration;.;destination_canonical_service=.=apigw;.;destination_canonical_revision=.=latest;.;request_protocol=.=http;.;response_flags=.=DC;.;connection_security_policy=.=none;.;response_code=.=0;.;grpc_response_status=.=;.;destination_cluster=.=Kubernetes;.;source_cluster=.=unknown;.;istio_request_duration_milliseconds: P0(nan,1000.0) P25(nan,59352.94117647059) P50(nan,59882.35294117647) P75(nan,60437.5) P90(nan,60775.0) P95(nan,60887.5) P99(nan,60977.5) P99.5(nan,60988.75) P99.9(nan,60997.75) P100(nan,61000.0)
reporter=.=destination;.;source_workload=.=unknown;.;source_workload_namespace=.=unknown;.;source_principal=.=unknown;.;source_app=.=unknown;.;source_version=.=unknown;.;source_canonical_service=.=unknown;.;source_canonical_revision=.=latest;.;destination_workload=.=apigw;.;destination_workload_namespace=.=project-apigw-integration;.;destination_principal=.=unknown;.;destination_app=.=apigw;.;destination_version=.=unknown;.;destination_service=.=apigw.project-apigw-integration.svc.cluster.local;.;destination_service_name=.=apigw;.;destination_service_namespace=.=project-apigw-integration;.;destination_canonical_service=.=apigw;.;destination_canonical_revision=.=latest;.;request_protocol=.=http;.;response_flags=.=DC;.;connection_security_policy=.=none;.;response_code=.=101;.;grpc_response_status=.=;.;destination_cluster=.=Kubernetes;.;source_cluster=.=unknown;.;istio_request_duration_milliseconds: P0(nan,64.0) P25(nan,64.5) P50(nan,65.0) P75(nan,1350000.0) P90(nan,1380000.0) P95(nan,1390000.0) P99(nan,1398000.0) P99.5(nan,1399000.0) P99.9(nan,1399800.0) P100(nan,1400000.0)
reporter=.=destination;.;source_workload=.=unknown;.;source_workload_namespace=.=unknown;.;source_principal=.=unknown;.;source_app=.=unknown;.;source_version=.=unknown;.;source_canonical_service=.=unknown;.;source_canonical_revision=.=latest;.;destination_workload=.=apigw;.;destination_workload_namespace=.=project-apigw-integration;.;destination_principal=.=unknown;.;destination_app=.=apigw;.;destination_version=.=unknown;.;destination_service=.=apigw.project-apigw-integration.svc.cluster.local;.;destination_service_name=.=apigw;.;destination_service_namespace=.=project-apigw-integration;.;destination_canonical_service=.=apigw;.;destination_canonical_revision=.=latest;.;request_protocol=.=http;.;response_flags=.=UC;.;connection_security_policy=.=none;.;response_code=.=503;.;grpc_response_status=.=;.;destination_cluster=.=Kubernetes;.;source_cluster=.=unknown;.;istio_request_duration_milliseconds: P0(nan,11000.0) P25(nan,33500.0) P50(nan,61000.0) P75(nan,82500.0) P90(nan,154000.00000000003) P95(nan,189000.0) P99(nan,205799.99999999997) P99.5(nan,207900.0) P99.9(nan,209579.99999999997) P100(nan,210000.0)

```

**Comment by bryan-kc** [comment]
Closing this issue as it is resolved.

Istio sidecar injection does not have requests/limits that are equal to each other. Our application was running on Guaranteed QoS due to strict requirements about having enough resources available for GC and other application resource lifecycle management.
https://kubernetes.io/docs/tasks/configure-pod-container/quality-service-pod/#create-a-pod-that-gets-assigned-a-qos-class-of-guaranteed

With the Istio sidecar enabled, the application switched to Burstable QoS without us being aware of it and thus caused the latencies seen here. The fix was to include the following manifest in our Deployment's specification so that our application remains with Guaranteed QoS.
```yaml
      initContainers: # The init container is defined to override the default injection template to ensure the Pod gets Guaranteed QoS
      - name: istio-init
        image: auto
        resources:
          limits:
            cpu: 1
            memory: 256Mi
          requests:
            cpu: 1
            memory: 256Mi
      containers:
      - name: istio-proxy # This Istio Proxy container is predefined here to override the default injection template to ensure the Pod gets Guaranteed QoS
        image: auto
        resources:
          limits:
            cpu: 1
            memory: 256Mi
          requests:
            cpu: 1
            memory: 256Mi
```

**Comment by bianpengyuan** [comment]
This is an interesting side effect... To help me understand, what happens is that because of changing of QoS class, your app pods are scheduled on node which does not have enough resources to handle burst CPU from GC, and thus causing the latency? I think this worth documented somewhere on istio.io.

**Comment by bryan-kc** [comment]
This goes a little beyond the scope of Istio sidecars, but that is the idea. We run a high throughput GraphQL backend service based on NodeJS. We found through performance tuning and testing that we needed dedicated cores to perform event loop and Garbage collection (https://v8.dev/blog/concurrent-marking). 

Thus prior to introducing Istio sidecars we created a manifest that gave our application Guaranteed QoS so that it had a dedicated core to perform on instead of several smaller cores
Example this application with guaranteed QoS
```shell
kubectl -n my-application-prod exec -it my-application-deadbeef123-xyz12 -c my-application -- /bin/sh -c 'grep Cpus_allowed /proc/self/status; grep . /sys/fs/cgroup/cpuset/cpuset.cpus /sys/fs/cgroup/cpuset/cpuset.cpu_exclusive'
Cpus_allowed:	0000,80000080
Cpus_allowed_list:	7,31
```

When automatic istio sidecar injection was enabled, the sidecars did not have container requests == limits, which means they will be scheduled with Burstable QoS and have several CPUs available. We believe this impacts GC and event loop lag due to having to switch between each CPU and compete for resources on a node.
```shell
kubectl -n my-application-prod exec -it my-application-deadbeef123-xyz12 -c my-application -- /bin/sh -c 'grep Cpus_allowed /proc/self/status; grep . /sys/fs/cgroup/cpuset/cpuset.cpus /sys/fs/cgroup/cpuset/cpuset.cpu_exclusive'
Cpus_allowed:	fff8,8ffff88f
Cpus_allowed_list:	0-3,7,11-27,31,35-47
/sys/fs/cgroup/cpuset/cpuset.cpus:0-3,7,11-27,31,35-47
/sys/fs/cgroup/cpuset/cpuset.cpu_exclusive:0
```

Because of the number and varying sizes of CPUs available we found this gets in the way of our application being able to perform GC and event loops. The performance degradation metrics are not due directly to Istio in this case, but our application suffering from being Burstable QoS and thus having latency issues.
